Forms: ensure custom CSS in global styles applies correctly to checkbox and radio wrappers

ramonjd opened the pull request:
## Proposed changes:

Resolves: https://github.com/Automattic/jetpack-roadmap/issues/2465

Base branch: https://github.com/Automattic/jetpack/pull/42890

This PR:

- Adds `wp-block-*` classes to field wrappers so that theme/user global styles apply.
- Ensures that block style classes (`is-style-*`) are applied at the wrapper level and not to the inner blocks so that any theme/user global style block style variation updates apply.
- Only applies the above changes to fields with inner blocks (presumably blocks that have been migrated according to https://github.com/Automattic/jetpack/pull/42890)

## Testing instructions:

1. In a post insert a form that contains some fields - include an email, multiple checkbox field and a single radio field. The email field is to test the theme.json snippet below.
2. Save. Make sure everything appears as expected on the frontend/editor.
3. Apply the "button" style to both fields.
4. Save. Make sure everything appears as expected on the frontend/editor.
5. Now update your theme's theme.json with some styles targeting these blocks (see JSON snippet below).
6. You guessed it... Make sure everything appears as expected on the frontend/editor.
7. Head over to the site editor and edit the custom CSS for the "button" variation. 
8. Wait for it.... Save. Make sure everything appears as expected on the frontend/editor.


```json
"blocks": {
	"jetpack/field-checkbox-multiple": {
	        "border": {
	                "width": "5px",
	                "radius": "2px",
	              	"color": "yellow"
	        },
	        "color": {
	                "background": "red"
	        },
	        "css": "font-size:12px",
	        "variations": {
	               "button": {
	 					"css": "background:purple;font-size:100px;"
	                 }
	        }
	},
	"jetpack/field-radio": {
	        "border": {
	                "width": "1px",
	                "radius": "12px",
	              	"color": "pink"
	        },
	        "color": {
	                "text": "blue"
	        },
	        "css": "font-size:19px",
	        "variations": {
	               "button": {
	 					"css": "background:red;font-size:11px;"
	                 }
	        }
	},
        "jetpack/field-email": {
                "border": {
                        "width": "1px",
                        "radius": "12px",
                        "color": "orange"
                },
                "color": {
                        "text": "white",
						"background": "grey" 
                },
				"spacing": {"padding":"10px"},
                "css": "font-size:1.3rem"
        },
}
```

| Editor | Frontend | Global style variation |
|--------|--------|--------|
| <img width="400" alt="Screenshot 2025-04-23 at 1 45 51 pm" src="https://github.com/user-attachments/assets/ee1ac7d5-18b2-4fa3-9910-4675f57f8584" /> | <img width="400" alt="Screenshot 2025-04-23 at 1 45 47 pm" src="https://github.com/user-attachments/assets/d7580280-e759-4256-a824-299b47294b7a" /> | <img width="400" alt="Screenshot 2025-04-23 at 1 45 01 pm" src="https://github.com/user-attachments/assets/5c100cb0-a221-4e01-abd4-fcd52ab68a92" /> |

ramonjd committed:
Refactor class attribute handling in Contact_Form_Field

- Trimmed type variable to ensure consistent class generation.
- Added specific classes for 'checkbox-multiple' and 'radio' types.
- Included TODOs for potential improvements and testing.

ramonjd committed:
changelog

ramonjd committed:
Add block styles for checkbox-multiple and radio fields

- Registered new block styles for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio'.
- Updated class handling in Contact_Form_Field to ensure proper application of styles.
- Adjusted CSS selectors to reflect changes in class structure for button styles.

aaronrobertshaw commented on projects/packages/forms/changelog/update-migrate-form-fields-to-inner-blocks-multiple-wrapper-styles:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: fixed
+
+Some form groups need block classes on their wrapper styles for global styles to apply
```
I think we can skip changelogs for minor stuff when merging into the #42890 branch. There should be one overarching changelog to cover the refactor to inner blocks.

ramonjd replied:
I've been conditioned to appease the linter.

aaronrobertshaw commented on projects/packages/forms/src/blocks/contact-form/class-contact-form-block.php:
```diff
@@ -257,6 +257,32 @@ public static function register_child_blocks() {
+		 * A follow up issue is to update them to use block.json files, which can be reused
+		 * in both JS and PHP block registration.
+		 */
+		register_block_style(
+			'jetpack/field-checkbox-multiple',
```
What's the minimum WP version for Jetpack? 

Recent versions should support passing multiple block names in an array. It's neither here nor there but would save the repeated `register_block_style` calls.

ramonjd replied:
Looks like the tests expect at least 6.6... so it should be find to use `string[]`. Thanks for the nudge.

ramonjd committed:
Enhance class handling for checkbox-multiple and radio fields

- Updated Contact_Form_Field to apply block style classes to wrapper divs for checkbox-multiple and radio types.
- Introduced a new method to extract block style classes from the class attribute.
- Modified tests to validate the addition of block style classes in the rendering process.

aaronrobertshaw commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -1474,27 +1475,53 @@ public function render_field( $type, $id, $label, $value, $class, $placeholder,
+		$block_classes   = '';
+		$wrap_classnames = $class . ' grunion-field'; // Classes for $shell_field_class wrapper.
+
+		/*
+		 * For the checkbox-multiple and radio fields, we need to add the `is-style-*` classes to the wrapper div
+		 * but not the input elements. This ensures any updates to block style in the theme
+		 * are applied to the wrapper div, not the input elements.
+		 */
```
It seems we're missing the `wp-block-jetpack-field-*` classes on wrappers. So if the custom CSS is only added to the block type e.g. Multiple Choice Field, the custom CSS selector won't match.

ramonjd replied:
Sorry, I don't follow. Do you mean `wp-block-jetpack-field-checkbox-multiple` and `wp-block-jetpack-field-radio` ? 

If so, they're added below and I see them in the frontend. There's also test coverage.

aaronrobertshaw replied:
I might have misread but those are only added when there are block style classes.

1. Add choice field
2. Go to site editor > Global Styles > Blocks > Multiple choice field
3. Expand advanced panel, add `background: red` to custom CSS
4. Save 
5. View on frontend

aaronrobertshaw replied:
<img width="1777" alt="Screenshot 2025-04-23 at 3 58 05 pm" src="https://github.com/user-attachments/assets/8d9d0220-0ac9-4627-b69a-a883825133a7" />

ramonjd replied:
Ah, no. Not you, me. I get it. 

I need to ensure that the `wp-block` classes are there even when a style hasn't been selected or is not available.

aaronrobertshaw replied:
Yeah, so I see this as just the problem scope being a little larger than initially thought.

What do you think of potentially avoiding the introduction of block class styles interfering with styles? In `block_attributes_to_shortcode_attributes` we could detect when we have the new inner blocks structure and update the shortcode's `class` attribute.

My thinking then is that this would inject the block class only when we want i.e. not for legacy forms.

aaronrobertshaw replied:
This will be a problem for more than just the choice fields. At a quick glance, I don't think any fields get their block classes on the frontend.

ramonjd replied:
> What do you think of potentially avoiding the introduction of block class styles interfering with styles? In block_attributes_to_shortcode_attributes we could detect when we have the new inner blocks structure and update the shortcode's class attribute.

I don't think I yet have enough context to be confident of anything. 

From what I have learned, `Contact_Form_Field::render_field` currently applies all members of `$class` to both the wrapper (as `-wrap`) and to the inner options (checkbox/radio). 

For `wp-block` and `is-style-*` classes we don't want that, so that method has to be short-circuited somehow.

> My thinking then is that this would inject the block class only when we want i.e. not for legacy forms.

Not saying it isn't possible to do the legwork in `block_attributes_to_shortcode_attributes`, but it might require some additional flag or something to tell `Contact_Form_Field::render_field` to treat the classes differently where the new inner blocks are detected.

I think. 😄 

> This will be a problem for more than just the choice fields. At a quick glance, I don't think any fields get their block classes on the frontend.

I'll look into that too. Won't hurt to widen the scope here to ensure whatever solution covers most bases. 

Thanks for the ideas!

aaronrobertshaw replied:
Good points. Thanks for setting me straight.

Another option, which might be a bit hacky depending on how you look at it, is to just add another shortcode attribute to store wrapper classes. Then map that to a property on the field class and use as desired.

aaronrobertshaw reviewed (commented):
Thanks for fixing up these oversights @ramonjd 🙇 

I'm still grokking the changes but hit an issue when testing custom CSS set in Global Styles for the field block types on the frontend. Is this something that is on the radar already?

ramonjd committed:
Refactor block style registration for checkbox-multiple and radio fields

- Consolidated block style registration for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio' into a single call.
- Updated comments to clarify the purpose of style registration and future improvements.
- Adjusted class handling in Contact_Form_Field to include new class naming conventions.

ramonjd committed:
Refactor block style application for checkbox-multiple and radio fields

- Streamlined the handling of block style classes for 'checkbox-multiple' and 'radio' types in Contact_Form_Field.
- Improved code readability by consolidating conditional checks for block style class assignment.
- Ensured consistent application of styles to wrapper divs for both field types.

ramonjd committed:
Enhance radio field rendering tests and block style application

- Added tests to validate rendering of radio fields with block style classes.
- Improved handling of block style classes for radio fields in Contact_Form_Field to ensure consistent application.
- Updated assertions in tests to reflect changes in class structure for radio fields.

ramonjd committed:
Remove outdated changelog entry for migrating form fields to inner blocks with multiple wrapper styles

ramonjd commented:
Thanks for the first pass.

> testing custom CSS set in Global Styles for the field block types on the frontend. Is this something that is on the radar already?

Quick answer, "no".

Long answer "not yet".

I can look into it. I wouldn't expect the solution to waiver too far from whatever the hell I'm doing here.

What's the definition of field block types here? 

Do you mean only the new inner blocks? 'jetpack/input', 'jetpack/label', `jetpack/option`? Basically the ones we can actually change in global styles (and not `jetpack/field-email` and others).

<img width="246" alt="Screenshot 2025-04-23 at 4 26 06 pm" src="https://github.com/user-attachments/assets/51fb76f5-39ed-4665-a83f-089d9e290486" />

ramonjd committed:
Enhance field wrapper class handling for contact form fields

- Introduced 'fieldwrapperclasses' attribute to allow custom wrapper classes for form fields.
- Updated rendering logic in Contact_Form_Field to utilize the new attribute for consistent styling.
- Refactored tests to validate the addition of 'fieldwrapperclasses' for various field types, ensuring proper application of block styles.

aaronrobertshaw commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -399,6 +435,20 @@ public static function block_attributes_to_shortcode_attributes( $atts, $type, $
 
 		// Process inner blocks to shortcode attributes.
 		if ( $block && ! empty( $block->parsed_block['innerBlocks'] ) ) {
+			/*
+			 * Add the `wp-block-jetpack-field-*` and `is-style-*` classes to the field wrapper div
+			 * This ensures any updates to field block styles in theme.json or global styles are
+			 * correctly applied.
+			 */
```
I believe prior to #42890, the choice fields used inner blocks for the inner options (just not the label as well). This might result in the `wp-block-jetpack-field-*` class being added for those legacy choice fields. 

This probably isn't a big deal. An alternative might be when processing the inner blocks for the new inner block types, flag the field block as needing the `wp-block` class and updating accordingly based on that.

ramonjd replied:
Thanks for flagging. I'll make a note to test next week 🙇🏻

ramonjd replied:
Got a draft going... updated tests to come

- https://github.com/Automattic/jetpack/pull/43268

aaronrobertshaw reviewed (approved):
I've given this one a quick smoke test and it is working as advertised for me.

There was one minor nit about when we inject the `wp-block` class but I'll leave it up to you to decide how far you wish to take it.

Other than that, I think this is fine to merge into the main PR in time for the main call for testing tomorrow.

andrewserong committed:
Try fixing a linting issue with the DataProvider

ramonjd commented:
Thanks a lot for testing folks! I'll push the button.
